independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,117 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.jar.JarFile;
+
+import io.smallrye.common.io.jar.JarFiles;
+
+public class JarFileReference {
+    // Guarded by an atomic reader counter that emulate the behaviour of a read/write lock.
+    // To enable virtual threads compatibility and avoid pinning it is not possible to use an explicit read/write lock
+    // because the jarFile access may happen inside a native call (for example triggered by the RunnerClassLoader)
+    // and then it is necessary to avoid blocking on it.
+    private final JarFile jarFile;
+
+    private final AtomicReference<JarFileReference> jarFileReference;
+
+    private final AtomicInteger readersCounter = new AtomicInteger(0);
+
+    private volatile boolean releasing = false;
+
+    private JarFileReference(AtomicReference<JarFileReference> jarFileReference, JarFile jarFile) {
+        this.jarFileReference = jarFileReference;
+        this.jarFile = jarFile;
+    }
+
+    /**
+     * Increase the readers counter of the jarFile and returns it.
+     *
+     * @return The jarFile wrapped by this reference or null if a release has been requested thus making the file no longer
+     *         accessible.
+     */
+    private JarFile readLockAcquire() {
+        readersCounter.incrementAndGet();
+        if (releasing) {
+            if (readersCounter.decrementAndGet() == 0) {
+                close();
```

Comment:
This the part which can break:
- T1 on release: set `release = true` and arrive till `readersCounter.get() == 0` check, not yet happening
- T2 has been able to read `releasing` and has decremented the read counter till 0
- T1 now check `readersCounter.get() == 0` as successfull too
- T2 and T1 both compete to perform the `close`

Reply:
This is true and I also found this possibility, anyway this shouldn't be a problem: at worst 2 threads will both call close() which however is idempotent.

Reply:
Calling close multiple times happens all the time at the HTTP layer for example and it's not a problem because as Mario mentions, it's idempotent.

All that to say that I don't believe this is an issue